engine-tests/src/test/java/org/terasology/config/flexible/FlexibleConfigTest.java

```diff
@@ -144,4 +150,108 @@ public void testSubscribedRemove() throws Exception {
             assertFalse(config.remove(id));
         }
     }
+
+    public static class JsonStorage {
+        private static final String CONFIG_AS_JSON = "{\n" +
+                "  \"engine-tests:TestSetting3\": {\n" +
+                "    \"a\": 31,\n" +
+                "    \"b\": \"string\",\n" +
+                "    \"c\": \"A2\"\n" +
+                "  },\n" +
+                "  \"engine-tests:TestSetting1\": \"A3\",\n" +
+                "  \"engine-tests:TestSetting2\": 21.0\n" +
+                "}";
+
+        private static final TestEnum TESTENUM_SETTING_VALUE = TestEnum.A3;
+        private static final double DOUBLE_SETTING_VALUE = 21.0;
+        private static final TestClass TESTCLASS_SETTING_VALUE = new TestClass();
+
+        private FlexibleConfig config;
+
+        private SettingImpl<TestEnum> testEnumSetting;
+        private SettingImpl<Double> doubleSetting;
+        private SettingImpl<TestClass> testClassSetting;
+
+        private void setupSettings() {
+            testEnumSetting = new SettingImpl<>(TestEnum.class, new SimpleUri("engine-tests:TestSetting1"), TestEnum.A1);
```

Comment:
From my perspective this way of instantiating a setting is quite a mouthful and I think from an immediateness and readability perspective we need like this:

`testEnumSetting = new IntegerSetting(new SimpleUri("engine-tests:TestSetting1"), TestEnum.A1);`

So, we need to reintroduce the concrete classes (*) and I'd like the SettingImpl class to be abstract.

(*) I.e.:
```
public class IntegerSetting extends SettingImpl {
    public IntegerSetting(SimpleUri id, Integer defaultValue) {
        super(Integer.class, id, defaultValue, null);
    }
}

```

Reply:
I was in a hurry when drafting this code. Obviously you'll add a new argument to this constructor (for the validator) and pass that to super instead of `null`.

Reply:
As discussed on slack, there is a way of getting rid of the `Class<T>` parameter, which will restore `SettingImpl` instantiation to how it was in the initial FC PR #2757 was merged:

```java
Setting<TestEnum> testEnumSetting = new SettingImpl<>(id1, TestEnum.A1);
Setting<Double> doubleSetting = new SettingImpl<>(id2, 30.0);
Setting<TestClass> testClassSetting = new SettingImpl<>(id3, null);
```

In contrast, making `SettingImpl` abstract and having concrete subclasses for each type would look like this:

```java
public class TestClassSetting extends SettingImpl<TestClass> {
    public TestClassSetting(SimpleUri id, TestClass defaultValue) {
        this(TestClass.class, id, defaultValue, null);
    }
        public TestClassSetting(SimpleUri id, TestClass defaultValue, 
                                SettingValueValidator<TestClass> validator) {
        super(TestClass.class, id, defaultValue, validator);
    }
}

// And so on for Integer, Boolean, Float, String, Vector3i, Side, etc.
```

 ```java
Setting<TestEnum> testEnumSetting = new TestEnumSetting(id1, TestEnum.A1);
Setting<Double> doubleSetting = new DoubleSetting(id2, 30.0);
Setting<TestClass> testClassSetting = new TestClassSetting(id3, null);
```

Reply:
On instantiation, would this work:

```Setting doubleSetting = new DoubleSetting(id2, 30.0);```

or would it generate an error? Any other consequence? Because that's the compactness I'd like to see when in use.

Reply:
@emanuele3d You'd have to specify the generic parameter, no matter what method is used: 
```java
Setting<Double> doubleSetting = new DoubleSetting(id2, 30.0);
```
or
```java
Setting<Double> doubleSetting = new SettingImpl<>(id2, 30.0);
```
or
```java
Setting<Double> doubleSetting = new SettingImpl<Double>(id2, 30.0);
```

Of course, if you do not want type information, you could levy type erasure and get rid of the type parameter, but then there would be no point in using generics. However, you may not need type information when simply adding a setting to a FC (which is the only place where setting instantiation will be used), in which case you can get rid of the type parameter.

To be fair, `Setting` instantiation will only be done _once_ by the code responsible for initializing the FlexibleConfig and the Settings. All future retrievals will be done by the `FlexibleConfig.get` method, so I personally do not think this matters too much. Also, I believe that all methods above exhibit similar amounts of readability.

Reply:
Indeed generics weren't exactly my top choice and if it was up to me I'd just rely on type erasure and settings consumers applying the appropriate cast on setting retrieval. 

But you know what, as I mentioned I'm a bit fed up with this whole discussion, so let's stick to the currently available form:

`Setting<Double> doubleSetting = new SettingImpl<>(id2, 30.0);`